clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -745,6 +745,10 @@ protected boolean onJoinPrepare(int generation, String memberId) {
         boolean onJoinPrepareAsyncCommitCompleted = false;
         // async commit offsets prior to rebalance if auto-commit enabled
         RequestFuture<Void> future = maybeAutoCommitOffsetsAsync();
+        // wait for commit offset response if future exist
+        if (future != null) {
+            client.poll(future, time.timer(rebalanceConfig.rebalanceTimeoutMs));
+        }
```

Comment:
If `client.poll(future, time.timer(rebalanceConfig.rebalanceTimeoutMs));` is still called here, the problem will recur, and the user will still block the `poll` method of kafkaConsumer. ` rebalanceConfig.rebalanceTimeoutMs` may be much larger than `pollDuration`

Suggest:
1. Promote the `future` variable in `onJoinPrepare` to the instance variable of `ConsumerCoordinator`. The variable name can be tentatively `rebalanceAutoCommitFuture`, and the initial value is `null`.
    ` private RequestFuture<Void> rebalanceAutoCommitFuture =null;`

2. Refactor the `onJoinPrepare` method. The `rebalanceAutoCommitFuture` can be completed after the user has called the `poll` method multiple times without blocking the user's `poll` method.
        ```
        boolean onJoinPrepareAsyncCommitCompleted = false;
        if(autoCommitEnabled && rebalanceAutoCommitFuture == null){
            // async commit offsets prior to rebalance if auto-commit enabled
            rebalanceAutoCommitFuture = maybeAutoCommitOffsetsAsync();
        }
        if (rebalanceAutoCommitFuture != null) {
            client.poll(rebalanceAutoCommitFuture, time.timer(0));
        }
        // return true when
        // 1. future is null, which means no commit request sent, so it is still considered completed
        // 2. offset commit completed
        // 3. offset commit failed with non-retriable exception
        if (rebalanceAutoCommitFuture == null)
            onJoinPrepareAsyncCommitCompleted = true;
        else if (rebalanceAutoCommitFuture.succeeded()) {
            onJoinPrepareAsyncCommitCompleted = true;
            rebalanceAutoCommitFuture = null;
        } else if (rebalanceAutoCommitFuture.failed() && !rebalanceAutoCommitFuture.isRetriable()) {
            log.error("Asynchronous auto-commit of offsets failed: {}", rebalanceAutoCommitFuture.exception().getMessage());
            onJoinPrepareAsyncCommitCompleted = true;
            rebalanceAutoCommitFuture = null;
        }
```

Reply:
@RivenSun2  Ah Yes, you are right.
i think in previous issue https://issues.apache.org/jira/browse/KAFKA-13310, the problem is that `maybeAutoCommitOffsetsSync` falls into an infinite loop if meet `UNKNOWN_TOPIC_OR_PARTITION` error.
But i think we need to wait for the commitOffset response and try to commit offset successfully before joinGroup, to  make 'no duplicate message' if every thing works well.

Your suggestion can work, but it will cause KafkaConsuemr.poll() to be in a busy loop between the time when commitOffsetRequest is in flight.

Do you think we can do:
* still wait for the response here
* limit the max retry time of commitOffset in onJoinPrepare function

Thanks

Reply:
1. It is not easy to find a suitable value for 'the max retry time'. The most important point is that we should ensure that the `poll` method of kafkaConsumer can always return within the `pollDuration` time.
2. I don't think it's too much of a problem to wait for the `rebalanceAutoCommitFuture` to complete by calling the poll method multiple times.
We can take advice from other experts.
Thanks.

Reply:
Thanks @RivenSun2 , I forgot the purpose of our previous change. 
The purpose of offset commit in `onJoinPrepare` is to make sure after rebalance, the other consumer can start from the correct offset. So, this offset commit must complete with the correct offset. In @RivenSun2 's suggestion, if we store the future as instance variable, and make sure there's only 1 inflight future, and that can fix the infinite offset commit issue in this PR. And since during this time, the group consumers are not able to consume messages due to rebalance in progress, we can make sure the offset commit is the only one and up-to-date one before reblancing. So I think this is a good solution.

For the comment:
>  it will cause KafkaConsuemr.poll() to be in a busy loop between the time when commitOffsetRequest is in flight.

I agree. I'm thinking we can backoff when `onJoinPrepare` return false like this:
```java
// return false when onJoinPrepare is waiting for committing offset
  if (!onJoinPrepare(generation.generationId, generation.memberId)) {
      needsJoinPrepare = true;

      // backoff some time  <-- new added
      timer.sleep(Math.min(timer.remainingMs(), rebalanceConfig.retryBackoffMs));

      //should not initiateJoinGroup if needsJoinPrepare still is true
      return false;
  }
```

WDYT? @RivenSun2 @aiquestion ?